### PR TITLE
Keras-api evaluation method return modification

### PR DESCRIPTION
**Pull request by Litchilitchy**
fix evaluation return to correspond with keras, created a new method `processEvaluateResultRDD`, not removing the original method

add a new `python` package in test module, now python-keras api could directly be tested under `PythonZooKerasSpec`

### CODE REVIEW

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala [hunk 1]
```diff
@@ -408,13 +408,33 @@ abstract class KerasNet[T](implicit val tag: ClassTag[T], implicit val ev: Tenso
...
    * @param batchSize Number of samples per batch.
    */
+
+  /*
   def evaluate(
       x: RDD[Sample[T]],
       batchSize: Int)
       (implicit ev: TensorNumeric[T]): Array[(ValidationResult, ValidationMethod[T])] = {
     require(this.vMethods != null, "Evaluation metrics haven't been set yet")
     this.evaluate(x, this.vMethods, Some(batchSize))
```
Comment (hkvision) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala]: Can we directly add Loss to vMethods here and call this.evaluate only once?

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala [hunk 2]
```diff
@@ -408,13 +408,33 @@ abstract class KerasNet[T](implicit val tag: ClassTag[T], implicit val ev: Tenso
...
   }
+  */
```
Comment (jason-dai) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala]: delete the codes if you don't need them

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala [hunk 3]
```diff
@@ -408,13 +408,33 @@ abstract class KerasNet[T](implicit val tag: ClassTag[T], implicit val ev: Tenso
...
+  def evaluate(
+      x: RDD[Sample[T]],
+      batchSize: Int)
+      (implicit ev: TensorNumeric[T]): Array[(ValidationResult, ValidationMethod[T])] = {
+    // require(this.vMethods != null, "Evaluation metrics haven't been set yet")
+
+    val loss = this.evaluate(x, Array(new Loss(this.criterion)), Some(batchSize))
```
Comment (jason-dai) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala]: just add loss to `this.vMethods` at compile time

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/python/PythonZooKeras.scala [hunk 4]
```diff
@@ -147,12 +147,19 @@ class PythonZooKeras[T: ClassTag](implicit ev: TensorNumeric[T]) extends PythonZ
     }.toList.asJava
   }
 
+  private def processEvaluateResultRDD(
```
Comment (jason-dai) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/python/PythonZooKeras.scala]: just `processEvaluateResult`, not RDD

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/python/PythonZooKeras.scala [hunk 5]
```diff
@@ -147,12 +147,19 @@ class PythonZooKeras[T: ClassTag](implicit ev: TensorNumeric[T]) extends PythonZ
...
+  }
+
   def zooEvaluate(
       module: KerasNet[T],
       x: JavaRDD[Sample],
-      batchSize: Int = 32): JList[EvaluatedResult] = {
+      batchSize: Int = 32): JList[Float] = {
     val resultArray = module.evaluate(toJSample(x), batchSize)
-    processEvaluateResult(resultArray)
+    processEvaluateResultRDD(resultArray)
```
Comment (jason-dai) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/python/PythonZooKeras.scala]: we should do this for both scala and python version of `evaluate`

Comment (zhichao-li) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/python/PythonZooKeras.scala]: You should change `processEvaluateResult` directly rather than creating a new one.

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala (hunk 3 above)
Comment (zhichao-li) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala]: Make you condition logic more clear. In this case, what you want to do is to assemble the correct vMethods  i.e
```
this.vMethods = if xxx else xxx
return this.evaluate(vMethods)
```

File: zoo/src/test/scala/com/intel/analytics/zoo/pipeline/api/keras/python/PythonZooKerasSpec.scala [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+import com.intel.analytics.bigdl.utils.Shape
+import com.intel.analytics.zoo.common.NNContext
+import com.intel.analytics.zoo.pipeline.api.keras.layers.{Dense, Flatten}
+import com.intel.analytics.zoo.pipeline.api.keras.models.Sequential
+import org.apache.spark.{SparkConf, SparkContext}
+import org.apache.spark.rdd.RDD
+import org.scalatest.{BeforeAndAfter, FlatSpec, Matchers}
+import com.intel.analytics.zoo.pipeline.api.keras.python.PythonZooKeras
+
+class PythonZooKerasSpec extends FlatSpec with Matchers with BeforeAndAfter {
```
Comment (hkvision) [inline zoo/src/test/scala/com/intel/analytics/zoo/pipeline/api/keras/python/PythonZooKerasSpec.scala]: put it into `TrainingSpec` or `TopologySpec`?

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala (hunk 1 above)
Comment (Litchilitchy) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala]: no, if we want to achieve we have to modify bigdl code

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala (hunk 3 above)
Comment (Litchilitchy) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala]: loss does not belong to subclass of ValidationMethods

**Comment by Litchilitchy** [comment]
@zhichao-li @jason-dai @hkvision I think you prefer modifying instead of adding the new one. I would just do the modification and would not keep the old code

File: zoo/src/test/scala/com/intel/analytics/zoo/pipeline/api/keras/python/PythonZooKerasSpec.scala (hunk 6 above)
Comment (Litchilitchy) [reply zoo/src/test/scala/com/intel/analytics/zoo/pipeline/api/keras/python/PythonZooKerasSpec.scala]: this is the first time to call python api and I think it would be more convenient if we just want to test python api?

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala (hunk 3 above)
Comment (Litchilitchy) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala]: @zhichao-li @hkvision and if we want to assemble them in once operation we have to make `loss` into `vMethod`? Is there any side-effect? I am not sure because `vMethod` is originally designed to take `ValidationMethod` subclass

Comment (Litchilitchy) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala]: @zhichao-li @jason-dai ok I think we could make it. Perhaps I remembered something wrong before

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala (hunk 1 above)
Comment (Litchilitchy) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala]: noted, we could make it

File: zoo/src/test/scala/com/intel/analytics/zoo/pipeline/api/keras/python/PythonZooKerasSpec.scala (hunk 6 above)
Comment (hkvision) [reply zoo/src/test/scala/com/intel/analytics/zoo/pipeline/api/keras/python/PythonZooKerasSpec.scala]: We don't need to test python in scala. Here we put unit test to cover the change in scala. And in python, you add another unit test probably in `test_simple_integration.py`.

Comment (Litchilitchy) [reply zoo/src/test/scala/com/intel/analytics/zoo/pipeline/api/keras/python/PythonZooKerasSpec.scala]: this could skip the `make-dist.sh` process and I think this is the advantage?

Comment (zhichao-li) [reply zoo/src/test/scala/com/intel/analytics/zoo/pipeline/api/keras/python/PythonZooKerasSpec.scala]: I think the more the better in respect of unitest if no side affect, but you should also add some python unitests to conver the python path.

Comment (hkvision) [reply zoo/src/test/scala/com/intel/analytics/zoo/pipeline/api/keras/python/PythonZooKerasSpec.scala]: I mean no need to create a separate file for this ut? as the function generateData seems already in TrainingSpec and you can put this ut there for convenience.

File: zoo/src/test/scala/com/intel/analytics/zoo/pipeline/api/keras/python/PythonZooKerasSpec.scala [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    val api = new PythonZooKeras[Float]()
+    val bigdlApi = sc.broadcast(new PythonBigDL[Float]())
+
+    /** python api require to take no type Sample
+      * and it takes JavaRDD as input
+      *
+      * use toPySample to convert to no type
+      * use toJavaRDD to convert to JavaRDD
+    **/
+    val jd = trainingData.map(j => bigdlApi.value.toPySample(j)).toJavaRDD()
```
Comment (hkvision) [inline zoo/src/test/scala/com/intel/analytics/zoo/pipeline/api/keras/python/PythonZooKerasSpec.scala]: It will be better to cover the ut in Python in this case. PythonZooKeras is not supposed to be tested in Scala.

File: zoo/src/test/scala/com/intel/analytics/zoo/pipeline/api/keras/python/PythonZooKerasSpec.scala (hunk 6 above)
Comment (Litchilitchy) [reply zoo/src/test/scala/com/intel/analytics/zoo/pipeline/api/keras/python/PythonZooKerasSpec.scala]: does function `generateData` only exist in `TrainingSpec`?

Comment (Litchilitchy) [reply zoo/src/test/scala/com/intel/analytics/zoo/pipeline/api/keras/python/PythonZooKerasSpec.scala]: @zhichao-li actually there is no new methods created in this pr. So hasn't any old test already exist for testing this pr?

Comment (Litchilitchy) [reply zoo/src/test/scala/com/intel/analytics/zoo/pipeline/api/keras/python/PythonZooKerasSpec.scala]: @hkvision modified
